### PR TITLE
Fix: Failover exit when option `-FS exit` is given

### PR DESCRIPTION
**Pull request by cypheron**
This addresses issues #425 (closed) #652 (closed) #672 (feature request). However, there could be other issues related to the mentioned bugs. For details see original comment.

### CODE REVIEW

**Comment by smurfy** [comment]
Well, this breaks all sorts of things, mainly ctrl+c shutdown.
I try to look into this issue and porting stratum client to poolmanager.
Poolmanager needs support for "exit" too, i wasn't aware of the feature.

**Comment by cypheron** [comment]
Before MinerAux.h `doFarm()` code has been refactored into pool manager etc. there existed a line of code stating `if (_remote == "exit") ..." in this very file. This line has been dropped/ignored in the refactoring process. StratumClient still contains the "exit" condition, but you guys still seem to be unaware of this feature, so it does not surprise me why it was broken at point. A lot of people used this feature for algo switching. Ethminer is basically not working anymore for these people. But apart from I am mentioning some implementation faults which need to be addressed.
Ctrl+c works fine for me, but that does not mean much. This needs to be tested throughly before anything is merged. In the first place I fixed the issue for myself because I was really pissed that ethminer would hang my mining process, even though a mechanism for failover exit existed in theory.
Refactoring `doStratum()` is a good addition, will clean up much of the messy code.

**Comment by smurfy** [comment]
@cypheron yea, sorry i did not understand the exit thing was a feature. But its quite use-full. so i will add it again. 

Since the new implementation of PoolManager actually retries a couple of times before switching the Pool, is this also ok with exit or, if fallback is exit it should exit as soon as the pool got a disconnect?  No retries of the main pool i mean.

**Comment by cypheron** [comment]
> Fallback is exit it should exit as soon as the pool got a disconnet

Yes, exactly! My suggestion is this: Inside `p_client->onDisconnected([&]()` only call `tryReconnect()` if active connection does not equal "exit".

File: ethminer/MinerAux.h [hunk 1]
```diff
@@ -951,7 +951,7 @@ class MinerCLI
 				client.reconnect();
 			});
 
-			while (g_running)
+			while (client.isRunning())
```
Comment (jean-m-cyr) [inline ethminer/MinerAux.h]: Here you are re-introducing a recently corrected bug.

File: libpoolprotocols/stratum/EthStratumClient.cpp [hunk 2]
```diff
@@ -56,6 +56,7 @@ EthStratumClient::EthStratumClient(Farm* f, MinerType m, string const & host, st
 
 EthStratumClient::~EthStratumClient()
 {
+	m_running = false;
```
Comment (jean-m-cyr) [inline libpoolprotocols/stratum/EthStratumClient.cpp]: This variable serves no viable purpose! It does not belong in the stratum client. If a client is instantiated, then it is by definition 'running'.

**Comment by cypheron** [comment]
@jean-m-cyr 
Just to quote your changes from #704 
> [...] In my view, if a pool protocol client is instantiated then
by definition it is running. [...] The loop now will never terminate, which is a problem
for controlled shutdown. [...]

You felt and are still feeling that `m_running` is unnecessary and removed it out of this feeling or out of personal preference. You also acknowledged that you created a loop (in MinerAux) that will now never exit. Do you see how this is working completely against the failover exit feature? You basically closed the window any exit possibility left open by #704.
We should not let personal feelings take over our judgement. The proposed changes in this request are not going to be merged anyway. smufy already suggested porting stratum code over to pool manager and include exit strategy in pool manager workloop. this should be the way to go and focus on. nonetheless for the failover exit to work the proposed changes in this request can freely be used as a reference.

**Comment by jean-m-cyr** [comment]
@cypheron Yes, I left the non ending loop knowing of smurpy's plans. That silly variable was being set on disconnect and causing the loop to terminate prematurely.

**Comment by smurfy** [comment]
I like/interested in the CUDAMiner, Farm  and will have a look regarding the possibility of that being a fix or improvement farm.stop   crash/not really working   problem.

I will look into ``-FS exit`` soon.

**Comment by jean-m-cyr** [comment]
I believe this has been superseded? Can it be closed?

**Comment by smurfy** [comment]
well this is still a todo, to implement cleanly

**Comment by jean-m-cyr** [comment]
Ah, ok. Though exit was in pool manager. My bad

**Comment by smurfy** [comment]
closing this, #828  is almost ready

**Pull request closed without merging**